### PR TITLE
ENH: Add separate numba kernels for groupby aggregations

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This PR adds separate groupby numba kernels, since the old shared kernels with Window required an expensive sort beforehand.

e.g. look at this snakeviz (pay attention to the ``take`` call. for reference, ``column_looper`` is the part doing the actual aggregation)

KPgbLhJsIxi4U9L7gyqBZ8TaJbgkSqg34gItFRdnMNZYnl_SFElErHYCdVtS9BD82yn2EHepAQ9P60-dbUtaXhSp3DSIe-_u8hOscAGnKFVBMCN0MKj-u7ccNK4f

Benchmarks
```
       before           after         ratio
     [43d12a23]       [6cf83167]
                      <groupby-numba-kernels>
-         197±5ms         66.6±1ms     0.34  groupby.GroupByNumbaAgg.time_frame_agg('float64', 'sum')
-         234±4ms       77.5±0.7ms     0.33  groupby.GroupByNumbaAgg.time_frame_agg('float64', 'var')
-         221±2ms       65.2±0.7ms     0.30  groupby.GroupByNumbaAgg.time_frame_agg('float64', 'mean')
-       20.1±0.1s         54.3±1ms     0.00  groupby.GroupByNumbaAgg.time_frame_agg('float64', 'max')
-       20.3±0.2s       54.3±0.8ms     0.00  groupby.GroupByNumbaAgg.time_frame_agg('float64', 'min')
```

Compared to Cython, we are still ~2x slower, but I think I can get back some more perf (would be a follow up PR).

If ``parallel=True`` is passed, though, we can be around ~2x faster than our Cython impl, though.

(This is on a Intel 2019 MBP, Intel core i7-9750H w/ 16 GB RAM)

cc @jbrockmendel if interested.
(You and I talked about parallelism in pandas a couple weeks back).

### CODE REVIEW

File: pandas/core/groupby/groupby.py
Comment (lithomas1) [inline pandas/core/groupby/groupby.py:1487]: Technically can remove this now, since all the existing numba kernels have been ported to have a groupby variant.

If you'd rather have me split this PR by kernel, this would be necessary (would remove after all the PRs land).

**Comment by mroeschke** [comment]
While there is a nice speedup here, I am a little concerned about proliferating implementations of reductions in the spirit of https://github.com/pandas-dev/pandas/issues/53261

File: pandas/core/groupby/groupby.py
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:1525]: not grouped_reduce?

File: pandas/core/_numba/kernels/var_.py
Comment (jbrockmendel) [inline pandas/core/_numba/kernels/var_.py:169]: these are probably np.intp?

**Comment by lithomas1** [comment]
> While there is a nice speedup here, I am a little concerned about proliferating implementations of reductions in the spirit of #53261

The problem is the numba kernels are basically useless without this.

I was doing a little benchmarking a little while back and numba was ~5x slower than Cython before this change.
Because the majority of time is spent in the sorting overhead, this means that even running the numba aggregations in parallel doesn't really provide a speedup.

With these kernels at least, I think we can share the Kahan summation part for ``sum``/``mean``/``var``. 
I could probably also reimplement ``mean`` in terms of sum, so the impact of this shouldn't be too horrible.

File: pandas/core/groupby/groupby.py
Comment (lithomas1) [reply pandas/core/groupby/groupby.py:1525]: It's possible I guess, but kwarg handling would be kinda nasty.

(``grouped_reduce`` doesn't take kwargs, so we'd need to the weird wrapper function thing ``_cython_agg_general`` does)

File: pandas/core/_numba/kernels/var_.py
Comment (lithomas1) [reply pandas/core/_numba/kernels/var_.py:169]: You mean do ``npt.NDArray[np.intp]``?

Comment (jbrockmendel) [reply pandas/core/_numba/kernels/var_.py:169]: yes

Comment (lithomas1) [reply pandas/core/_numba/kernels/var_.py:169]: Done.

**Comment by lithomas1** [comment]
Gentle ping 
@jbrockmendel @mroeschke @rhshadrach

File: pyright_reportGeneralTypeIssues.json
Comment (jbrockmendel) [inline pyright_reportGeneralTypeIssues.json:21]: is there any prospect of this being fixed in numba?  i.e. condition under which we can remove this?

**Comment by jbrockmendel** [comment]
> If parallel=True is passed, though, we can be around ~2x faster than our Cython impl, though.

Is this one of the use cases where there is a prospect of parallelizing the cython version too?

**Comment by lithomas1** [comment]
> Is this one of the use cases where there is a prospect of parallelizing the cython version too?

In the near future, no, since shipping openmp in wheels is a PITA. 
I think we need an openmp runtime package/wheel or something to depend on on PyPI before moving forwards with parallelizing Cython.

numba seems to at least have figured out how to solve this problem partially. I think you can use the tbb backend if installing with pip and on conda-forge you can choose between tbb/openmp.

File: pyright_reportGeneralTypeIssues.json
Comment (lithomas1) [reply pyright_reportGeneralTypeIssues.json:21]: https://github.com/numba/numba/issues/5474

Comment (jbrockmendel) [reply pyright_reportGeneralTypeIssues.json:21]: can you add something like "# TODO(numba#5474): can be removed once numba#5474 is addressed"

**Comment by jbrockmendel** [comment]
no objection in principle, will defer to matt since he's more involved in manitaining this code

File: pandas/core/_numba/kernels/mean_.py
Comment (lithomas1) [reply pandas/core/_numba/kernels/mean_.py]: I don't think so, my bad. I copied it from the sliding window code.

This is updated now.

File: pyright_reportGeneralTypeIssues.json
Comment (lithomas1) [reply pyright_reportGeneralTypeIssues.json]: Looks like we lost the ability to add comments in the pyright json config file.

**Comment by lithomas1** [comment]
@mroeschke 
Do you have time to take a look at this soon?

(I'd like to get this in, if not the apply stuff, at least for 2.1)

**Review by mroeschke — APPROVED** [review]
LGTM, but still hoping one day we can still have one aggregation kernel that can apply to both window and groupby (and standard) functions 🤞

**Comment by mroeschke** [comment]
Thanks @lithomas1

**Pull request merged**
